IdentityCore/src/util/MSIDKeyOperationUtil.m

```diff
@@ -0,0 +1,130 @@
+    }
+    return isSupported;
+}
+
+- (MSIDJwtAlgorithm)getJwtAlgorithmForKey:(SecKeyRef)privateKey context:(id<MSIDRequestContext> _Nullable)context error:(NSError * _Nullable __autoreleasing * _Nullable)error
+{
+    if (privateKey == NULL)
+    {
+        [self generateErrorWithMessage:@"No key to determine signing algorithm" underlyingError:nil context:context error:error];
+        return nil;
+    }
+    
+    NSError *ecdsaAlgError;
+    if ([[MSIDKeyOperationUtil sharedInstance] isOperationSupportedByKey:kSecKeyOperationTypeSign algorithm:kSecKeyAlgorithmECDSASignatureDigestX962SHA256 key:privateKey context:context error:&ecdsaAlgError])
+    {
+        return MSID_JWT_ALG_ES256;
+    }
+
+    if ([[MSIDKeyOperationUtil sharedInstance] isOperationSupportedByKey:kSecKeyOperationTypeSign algorithm:kSecKeyAlgorithmRSASignatureDigestPKCS1v15SHA256 key:privateKey context:context error:error])
+    {
+        return MSID_JWT_ALG_RS256;
+    }
+    
+    if (error && ecdsaAlgError)
+    {
+        MSID_LOG_WITH_CTX(MSIDLogLevelError, context, @"Key does not support signing any supported algorithms or key is public key");
+    }
+    return nil;
+}
+
+- (NSData *)getSignatureForDataWithKey:(NSData *)rawData privateKey:(SecKeyRef)privateKey context:(id<MSIDRequestContext>)context error:(NSError * _Nullable __autoreleasing *)error
+{
+    if (!rawData)
+    {
+        [self generateErrorWithMessage: @"No data to sign" underlyingError:nil context:context error:error];
+        return nil;
+    }
+    
+    if (privateKey == NULL)
+    {
+        [self generateErrorWithMessage:@"No key defined to sign data with" underlyingError:nil context:context error:error];
+        return nil;
+    }
+    
+    SecKeyAlgorithm algorithm = nil;
+    // Since Secure enclave only supports ECC NIST P-256 curve key we can assume key is used for ECDSA
+    if ([[MSIDKeyOperationUtil sharedInstance] isKeyFromSecureEnclave:privateKey] && [[MSIDKeyOperationUtil sharedInstance] isOperationSupportedByKey:kSecKeyOperationTypeSign algorithm:kSecKeyAlgorithmECDSASignatureMessageX962SHA256 key:privateKey context:context error:error])
+    {
+        algorithm = kSecKeyAlgorithmECDSASignatureMessageX962SHA256;
+    }
+    else if ([[MSIDKeyOperationUtil sharedInstance] isOperationSupportedByKey:kSecKeyOperationTypeSign algorithm:kSecKeyAlgorithmRSASignatureMessagePKCS1v15SHA256 key:privateKey context:context error:error])
+    {
+        algorithm = kSecKeyAlgorithmRSASignatureMessagePKCS1v15SHA256;
+    }
+    
+    if (algorithm == NULL)
+    {
+        [self generateErrorWithMessage:@"Signing algorithm could not be determined for supplied key because key is not of supported type or it is a public key." underlyingError:nil context:context error:error];
+        return nil;
+    }
+    CFErrorRef *subError = NULL;
+    NSData *signature = (NSData *)CFBridgingRelease(SecKeyCreateSignature(privateKey,
```

Comment:
Are you sure that CFBridgingRelease(null) won't crash? (I'm  just double checking, since I don't know)

Reply:
CFRelease(NULL) causes a crash but CFBridgingRelease(NULL) doesn't crash. I followed the Apple docs example to create signature and they have the same usage : https://developer.apple.com/documentation/security/certificate_key_and_trust_services/keys/signing_and_verifying?language=objc. Tried CFBridgingRelease(NULL) in a test case to see if it crashes and it doesn't